csrc/preseg_passes/move_gather.h

```diff
@@ -0,0 +1,75 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <preseg_passes/optimization_pass.h>
+
+namespace nvfuser::preseg_passes {
+
+// A pre-segmenter pass that moves gather operations ahead of producer
+// unary pointwise ops such as cast. We move take_along_axis ahead of another op
+// if that op is a unary pointwise op such as cast or neg, or it's a squeeze op.
+// Following 2 examples demonstrated what this pass does:
+
```

Comment:
BTW, if this is a case with a downcast `cast`, by moving the gather up, we are going to inflate temporary buffer at the segment.

i.e. the memory save only works when we can push `Gather` to an existing segment boundary.